### PR TITLE
add new basic:label unittest file

### PR DESCRIPTION
**Pull request by devdekunle**
please @bluca @poettering I created a pull request before for this same task, I am having a challenge with the testing of each function in the src/basic/label.c file. The file has a static global struct LabelOps that is set through the label_ops_set function. After testing the label_ops_set function in a TEST block, i try to test the other two functions, label_ops_pre and label_ops_post. i try to run each test in separate TEST blocks but the initial value of the global label_ops is always set because the label_ops_set only allows the struct to be set just once hence making it impossible for me to start each new TEST block with fresh values and my tests always fails because label_ops in the label.c file can only be set once. Please kindly guide me

### CODE REVIEW

File: src/basic/label.c [hunk 1]
```diff
@@ -10,7 +10,6 @@ static const LabelOps *label_ops = NULL;
 int label_ops_set(const LabelOps *ops) {
         if (label_ops)
                 return -EBUSY;
-
```
Comment (bluca) [inline src/basic/label.c]: spurious change

File: src/test/test-label.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+#include "fs-util.h"
+#include "label.h"
+#include "path-util.h"
+#include "string-util.h"
+#include "tests.h"
+
+static int check_path(int dir_fd, const char *path) {
+
+        assert(dir_fd >= 0);
+        assert(path);
```
Comment (bluca) [inline src/test/test-label.c:16]: add new line after the asserts, not before, here and elsewhere

File: src/test/test-label.c [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        if (!is_path(path) || !path_is_valid(path) || isempty(path))
+                return -EINVAL;
+
+        if (strlen(path) > 40)
+                return -ENAMETOOLONG;
+
+        if (!path_is_safe(path))
+                return -ENOTDIR;
+
+        /*assume a case where a specific label isn't allowed*/
```
Comment (bluca) [inline src/test/test-label.c]: missing spaces after `/*` and before `*/`, here and elsewhere

File: src/test/test-label.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        if (path_equal(path, "/restricted_directory"))
+                return -EACCES;
+        return 0;
+}
+static int pre_labelling_func(int dir_fd, const char *path, mode_t mode) {
+        int ret;
```
Comment (bluca) [inline src/test/test-label.c]: use `r` not `ret`, here and elsewhere

File: src/test/test-label.c [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+static int post_labelling_func(int dir_fd, const char *path) {
+        /*sample post_labelling task*/
+       int ret;
+
+        /*assume label policies that restrict certain labels*/
+
+        ret = check_path(dir_fd, path);
+        if (ret < 0) {
```
Comment (bluca) [inline src/test/test-label.c]: no brackets around single lines ifs, here and elsewhere

File: src/test/test-label.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        /* custom post labelling logic */
+        return 0; /*on sucess*/
+}
+
+TEST(label_ops) {
+
+        static const LabelOps test_label_ops = {
+                .pre = pre_labelling_func,
+                .post = post_labelling_func,
+        };
```
Comment (bluca) [inline src/test/test-label.c:62]: new line after variable declarations, not before

**Review by bluca — CHANGES_REQUESTED** [review]

**Comment by bluca** [comment]
> The file has a static global struct LabelOps that is set through the label_ops_set function. After testing the label_ops_set function in a TEST block, i try to test the other two functions, label_ops_pre and label_ops_post. i try to run each test in separate TEST blocks but the initial value of the global label_ops is always set because the label_ops_set only allows the struct to be set just once hence making it impossible for me to start each new TEST block with fresh values and my tests always fails because label_ops in the label.c file can only be set once. Please kindly guide me

Add a new function, in a separate commit first, `label_ops_reset` or so, that sets it back to NULL, and use it from the tests

**Comment by bluca** [comment]
rebase and squash fixup commits

**Comment by devdekunle** [comment]
> > The file has a static global struct LabelOps that is set through the label_ops_set function. After testing the label_ops_set function in a TEST block, i try to test the other two functions, label_ops_pre and label_ops_post. i try to run each test in separate TEST blocks but the initial value of the global label_ops is always set because the label_ops_set only allows the struct to be set just once hence making it impossible for me to start each new TEST block with fresh values and my tests always fails because label_ops in the label.c file can only be set once. Please kindly guide me
> 
> Add a new function, in a separate commit first, `label_ops_reset` or so, that sets it back to NULL, and use it from the tests

Thank you very much for your response @bluca do you mean i should add the function to reset label_ops in the basic:label.c file, the use it in the test:label.c file?

**Comment by bluca** [comment]
yes

**Comment by devdekunle** [comment]
> yes

Thank you @bluca I have implemented the changes, kindly review

File: src/test/test-label.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        assert_se(label_ops_pre(2, "abcd", 0644) == -EINVAL);
+        assert_se(label_ops_pre(1, "/wekrgoierhgoierhqgherhgwklegnlweehgorwfkryrit", 0755) == -ENAMETOOLONG);
+}
+
+TEST(label_ops_post) {
+
+        static const LabelOps test_label_ops = {
+                .pre = NULL,
+                .post = post_labelling_func,
+        };
```
Comment (bluca) [inline src/test/test-label.c:92]: add new line after this, remove from above

**Comment by bluca** [comment]
rebase and squash the fixup commit, and adjust the commit message, no need to list all the individual incremental changes made in the PR

**Pull request closed without merging**